azurelinuxagent/ga/update.py

```diff
@@ -912,40 +923,101 @@ def _check_and_download_agent_if_upgrade_available(self, protocol, base_version=
+                return False
+        else:
+            # If no requested version specified in the Goal State, follow the old auto-update logic
+            # Note: If the first Goal State contains a requested version, this timer won't start (i.e. self.last_attempt_time won't be updated).
+            # If any subsequent goal state does not contain requested version, this timer will start then, and we will
+            # download all versions available in PIR and auto-update to the highest available version on that goal state.
+            now = time.time()
+            if self.last_attempt_time is not None:
+                next_attempt_time = self.last_attempt_time + conf.get_autoupdate_frequency()
+            else:
+                next_attempt_time = now
+            if next_attempt_time > now:
+                return False
 
-            # Set the agents to those available for download at least as
-            # current as the existing agent and remove from disk any agent
-            # no longer reported to the VM.
-            # Note:
-            #  The code leaves on disk available, but blacklisted, agents
-            #  so as to preserve the state. Otherwise, those agents could be
-            #  again downloaded and inappropriately retried.
+            logger.info("No requested version specified, checking for all versions for agent update (family: {0})",
+                        family)
+
+            self.last_attempt_time = now
+
+        try:
+            # If we make it to this point, then either there is a requested version in a new GS (new auto-update model),
+            # or the 1hr time limit has elapsed for us to check the agent manifest for updates (old auto-update model).
+            pkg_list = ExtHandlerPackageList()
+
+            # If the requested version is the current version, don't download anything;
+            #       the call to purge() below will delete all other agents from disk
+            # In this case, no need to even fetch the GA family manifest as we don't need to download any agent.
+            if requested_version is not None and requested_version == CURRENT_VERSION:
+                packages_to_download = []
+                logger.info("The requested version is running as the current version: {0}".format(requested_version))
+            else:
+                pkg_list = protocol.get_vmagent_pkgs(manifests[0])
+                packages_to_download = pkg_list.versions
+
+            # Verify the requested version is in GA family manifest (if specified)
+            if requested_version is not None and requested_version != CURRENT_VERSION:
+                for pkg in pkg_list.versions:
+                    if FlexibleVersion(pkg.version) == requested_version:
+                        # Found a matching package, only download that one
+                        packages_to_download = [pkg]
+                        break
+
+                if packages_to_download == pkg_list.versions:
```

Comment:
Is it possible to have only requested version in `pkg_list` (manifest)?

Reply:
That's a good point. Its not possible generally, but still a possibility. Let me just use a flag instead

Reply:
Done